src/OpenLoco/src/GameCommands/RemoveTown.cpp

```diff
@@ -0,0 +1,161 @@
+#include "Audio/Audio.h"
+#include "Economy/Expenditures.h"
+#include "GameCommands.h"
+#include "Localisation/StringIds.h"
+#include "Localisation/StringManager.h"
+#include "Map/BuildingElement.h"
+#include "Map/RoadElement.h"
+#include "Map/TileElement.h"
+#include "Map/TileLoop.hpp"
+#include "Map/TileManager.h"
+#include "MessageManager.h"
+#include "Objects/ObjectManager.h"
+#include "S5/S5.h"
+#include "Ui/WindowManager.h"
+#include "ViewportManager.h"
+#include "World/IndustryManager.h"
+#include "World/StationManager.h"
+#include "World/TownManager.h"
+
+using namespace OpenLoco::Interop;
+using namespace OpenLoco::World;
+
+namespace OpenLoco::GameCommands
+{
+    // 0x0049711F
+    static uint32_t removeTown(const TownRemovalArgs& args, const uint8_t flags)
+    {
+        for (auto& station : StationManager::stations())
+        {
+            if (station.town == args.townId)
+            {
+                setErrorText(StringIds::all_stations_near_this_town_must_be_removed_first);
+                return FAILURE;
+            }
+        }
+
+        if (!(flags & Flags::apply))
+        {
+            return 0;
+        }
+
+        // NB: vanilla did not set an expenditute type
+        GameCommands::setExpenditureType(ExpenditureType::Construction);
+
+        // Iterate over the entire map to find town tiles
+        // TODO: can't we do better than vanilla for this? e.g. a radius around the town centre?
+        TilePosRangeView tileLoop{ { 1, 1 }, { kMapColumns - 1, kMapRows - 1 } };
+
+        for (auto& tilePos : tileLoop)
+        {
+            auto tile = TileManager::get(tilePos);
+
+            for (auto& element : tile)
```

Comment:
We will need to restart this tile element loop after any removal (building or road). Or you need to store all the removals for them to be performed at the end. I've used a variety of either throughout. This is due to the pointer invalidation after a removal. (It could also be avoided by iterating the tile in reverse order but lets not make this more complex).

Reply:
Could you point me to how best to reset the tile iterator?

Reply:
the simple way is: 

bool resetLoop = true;
while(resetLoop){
    resetLoop = false;
    auto tile = TileManager::get(tilePos);
    for (auto& element : tile)
    {
        // do the for loop
        // in for loop after a removal do
            resetLoop = true;
            break;
    }
}